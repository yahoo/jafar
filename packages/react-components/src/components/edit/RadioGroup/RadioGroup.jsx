/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import PropTypes from 'prop-types';
import Radio from '@material-ui/core/Radio';
import FormControlLabel from '@material-ui/core/FormControlLabel';
import MuiRadioGroup from '@material-ui/core/RadioGroup';

/**
 * Represent a single select of any value
 * 
 * Import <a target="_blank" 
 href="https://github.com/yahoo/jafar/blob/master/packages/react-components/src/components/edit/RadioGroup/RadioGroup.jsx">
 RadioGroup</a> from '@jafar-org/react-components/edit/RadioGroup'
 */
export default class RadioGroup extends React.Component {
  static propTypes = {
    value: PropTypes.any,
    state: PropTypes.shape({
      items: PropTypes.arrayOf(PropTypes.shape({
        label: PropTypes.string.isRequired,
        value: PropTypes.any.isRequired,
        disabled: PropTypes.bool,
      })).isRequired,
      inline: PropTypes.bool,
    }),
    disabled: PropTypes.bool,
    onValueChange: PropTypes.func.isRequired,
  };

  static defaultProps = {
    value: '',
    state: {
      items: [],
      inline: false,
    },
    disabled: false,
  };

  render() {
    return (
      <MuiRadioGroup value={this.props.value} onChange={this.onChange} row={this.props.state.inline}>
        {this.props.state.items.map((item) => {
          return (
            <FormControlLabel
              key={item.value} value={item.value} label={item.label} 
              disabled={this.props.disabled || item.disabled} control={<Radio />}
            />);
        })}
      </MuiRadioGroup>
    );
  }

  onChange = (e) => {
    this.props.onValueChange(e.target.value);
  }
}
