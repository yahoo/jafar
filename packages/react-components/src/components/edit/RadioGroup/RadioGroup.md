<h3>Examples</h3>

<h4>Usage in jafar form</h4>

```javascript
const Form = require('@jafar/react-form/Form').default;
const Field = require('@jafar/react-form/Field').default;

const model = {
  id: 'simple',
  fields: {
    bestFriend: {
      label: 'Best Friend',
      path: 'bestFriend',
      component: {
        name: 'RadioGroup',
        state: {
          items: [{
              value: '1',
              label: 'Ross Geller'
          }, {
              value: '2',
              label: 'Monica Geller'
          }, {
              value: '3',
              label: 'Rachel Green'
          }, {
              value: '4',
              label: 'Chandler Bing'
          }, {
              value: '5',
              label: 'Joey Tribbiani'
          }, {
              value: '6',
              label: 'Phoebe Buffay'
          }],
        },
      },
      required: true,
      disableTerm: {
          name: 'disableMe',
      },
      excludeTerm: {
          name: 'excludeMe',
      },
    },
  },
  data: {
    bestFriend: '2', // remove to see initial value undefined, and required error
  }
};

const resources = {
  components: { 
    RadioGroup: {
      renderer: RadioGroup,
    },
  },
  terms: {
      disableMe: { func: () => false }, // convert to true to see the field disabled
      excludeMe: { func: () => false }, // convert to true to see the field excluded
  },
};

<Form model={model} resources={resources}>
    <Field id='bestFriend' />
</Form>
```

<h4>Simple</h4>

```javascript
initialState = { 
  value: '2',
  state: {
    items: [{
            value: '1',
            label: 'Ross Geller'
        }, {
            value: '2',
            label: 'Monica Geller'
        }, {
            value: '3',
            label: 'Rachel Green'
        }, {
            value: '4',
            label: 'Chandler Bing'
        }, {
            value: '5',
            label: 'Joey Tribbiani'
        }, {
            value: '6',
            label: 'Phoebe Buffay'
        }],
  },
};

 <RadioGroup
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
  />
```

<h4>Inline</h4>

```javascript
initialState = { 
  value: '4',
  state: {
    inline: true,
    items: [{
            value: '1',
            label: 'Ross Geller'
        }, {
            value: '2',
            label: 'Monica Geller'
        }, {
            value: '3',
            label: 'Rachel Green'
        }, {
            value: '4',
            label: 'Chandler Bing'
        }, {
            value: '5',
            label: 'Joey Tribbiani'
        }, {
            value: '6',
            label: 'Phoebe Buffay'
        }],
  },
};
 <RadioGroup
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
  />
```

<h4>Initial value undefined</h4>

```javascript
initialState = { 
  value: undefined,
  state: {
    items: [{
        value: '1',
        label: 'Ross Geller'
    }, {
        value: '2',
        label: 'Monica Geller'
    }, {
        value: '3',
        label: 'Rachel Green'
    }, {
        value: '4',
        label: 'Chandler Bing'
    }, {
        value: '5',
        label: 'Joey Tribbiani'
    }, {
        value: '6',
        label: 'Phoebe Buffay'
    }],
  },
};
 <RadioGroup
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
  />
```

<h4>Disabled</h4>

```javascript
initialState = { 
  value: '4',
  state: {
    items: [{
            value: '1',
            label: 'Ross Geller'
        }, {
            value: '2',
            label: 'Monica Geller'
        }, {
            value: '3',
            label: 'Rachel Green'
        }, {
            value: '4',
            label: 'Chandler Bing'
        }, {
            value: '5',
            label: 'Joey Tribbiani'
        }, {
            value: '6',
            label: 'Phoebe Buffay'
        }],
  },
};
 <RadioGroup
  value={state.value}
  state={state.state}
  disabled={true}
  onValueChange={(value) => {
    setState({ value });
  }}
  />
```

<h4>Specific Items disabled</h4>

```javascript
initialState = { 
  value: '4',
  state: {
    items: [{
            value: '1',
            label: 'Ross Geller'
        }, {
            value: '2',
            label: 'Monica Geller',
            disabled: true
        }, {
            value: '3',
            label: 'Rachel Green'
        }, {
            value: '4',
            label: 'Chandler Bing'
        }, {
            value: '5',
            label: 'Joey Tribbiani'
        }, {
            value: '6',
            label: 'Phoebe Buffay',
            disabled: true
        }],
  },
};

<RadioGroup
  value={state.value}
  state={state.state}
  onValueChange={(value) => {
    setState({ value });
  }}
/>
```