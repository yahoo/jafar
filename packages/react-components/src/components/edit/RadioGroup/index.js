/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import RadioGroup from './RadioGroup';

export default RadioGroup;
