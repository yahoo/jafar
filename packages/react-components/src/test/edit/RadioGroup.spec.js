/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { shallow } from 'enzyme';
import RadioGroup from '../../components/edit/RadioGroup/index';

describe('<RadioGroup />', () => {
  let component;
  let onValueChangeSpy;
  const value = 'value2';
  const state = {
    items: [
      { label: 'label 1', value: 'value1' },
      { label: 'label 2', value: 'value2' },
      { label: 'label 3', value: 'value3' }],
  };
  const disabled = false;
  const invalid = false;

  beforeEach(() => {
    onValueChangeSpy = jest.fn();
  });

  it('should render provided data', () => {
    component = shallow(
      getComponent(value, state, disabled, invalid, onValueChangeSpy)
    );
    expect(component).toMatchSnapshot();
  });

  function getComponent(value, state, disabled, invalid, onValueChange) {
    return (
      <RadioGroup
        value={value}
        state={state}
        disabled={disabled}
        invalid={invalid}
        onValueChange={onValueChange}
      />
    );
  }
});
