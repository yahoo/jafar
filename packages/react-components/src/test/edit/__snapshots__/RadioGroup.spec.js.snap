// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RadioGroup /> should render provided data 1`] = `
<ForwardRef(RadioGroup)
  onChange={[Function]}
  value="value2"
>
  <WithStyles(ForwardRef(FormControlLabel))
    control={<ForwardRef(WithStyles) />}
    key="value1"
    label="label 1"
    value="value1"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    control={<ForwardRef(WithStyles) />}
    key="value2"
    label="label 2"
    value="value2"
  />
  <WithStyles(ForwardRef(FormControlLabel))
    control={<ForwardRef(WithStyles) />}
    key="value3"
    label="label 3"
    value="value3"
  />
</ForwardRef(RadioGroup)>
`;
